app/controllers/forest_liana/sessions_controller.rb

```diff
@@ -1,108 +1,80 @@
+        two_factor_registration: two_factor_registration,
+        two_factor_token: two_factor_token,
+      )
     end
 
-    def create_with_google
+    def login_with_google
       @error_message = nil
 
-      rendering_id = params['renderingId']
       forest_token = params['forestToken']
+      rendering_id = params['renderingId']
+      project_id = params['projectId']
+      two_factor_token = params['token']
+      two_factor_registration = params['twoFactorRegistration']
 
-      user = GoogleAuthorizedUserGetter.new(rendering_id, forest_token).perform()
-      token = encode_token(user) if user
-
-      if token
-        render json: { token: token }, serializer: nil
-      else
-        if @error_message
-          render serializer: nil, json: JSONAPI::Serializer.serialize_errors(
-            [{ detail: @error_message }]), status: :unauthorized
-        else
-          head :unauthorized
-        end
-      end
+      process_login(
+        use_google_authentication: true,
+        rendering_id: rendering_id,
+        project_id: project_id,
+        auth_data: { :forest_token => forest_token },
+        two_factor_registration: two_factor_registration,
+        two_factor_token: two_factor_token,
+      )
     end
 
     private
 
-    def check_user
-      if has_internal_authentication?
-        # NOTICE: Use the ForestUser table for authentication.
-        user = @user_class.find_by(email: params['email'])
-        user if !user.blank? && authenticate_internal_user(user['password_digest'])
-      else
-        # NOTICE: Query Forest server for authentication.
-        fetch_allowed_users
-
-        ForestLiana.allowed_users.find do |allowed_user|
-          allowed_user['email'] == params['email'] &&
-            BCrypt::Password.new(allowed_user['password']) == params['password']
+    def process_login(
+      use_google_authentication:,
+      rendering_id:,
+      project_id:,
+      auth_data:,
+      two_factor_registration:,
+      two_factor_token:
+    )
+      begin
+        if two_factor_registration && two_factor_token.nil?
+          raise ForestLiana::Errors::HTTP401Error
         end
-      end
-    end
 
-    def fetch_allowed_users
-      AllowedUsersGetter.new(params['renderingId']).perform()
-    end
+        # TODO: Add ip whitelist retrieving when it exists.
 
-    def has_internal_authentication?
-      @user_class && defined? @user_class
-    end
+        reponse_data = LoginHandler.new(
+          rendering_id,
+          auth_data,
+          use_google_authentication,
+          two_factor_registration,
+          project_id,
+          two_factor_token
+        ).perform
 
-    def authenticate_internal_user(password_digest)
-      BCrypt::Password.new(password_digest).is_password?(params['password'])
-    end
+      rescue ForestLiana::Errors::ExpectedError => error
+        error.display_error
+        error_data = JSONAPI::Serializer.serialize_errors([{ detail: error.message }])
```

Comment:
Only the higher exception message is displayed, so we just need to catch and rethrow to hide a sensitive exception message.